pilot/pkg/config/kube/ior/route.go

```diff
@@ -627,3 +367,155 @@ func hostHash(name string) string {
 	hash := sha256.Sum256([]byte(name))
 	return hex.EncodeToString(hash[:8])
 }
+
+func (r *route) reconcileGateway(config *config.Config, routes *v1.RouteList) error {
+	gateway, ok := config.Spec.(*networking.Gateway)
+
+	if !ok {
+		return fmt.Errorf("could not decode spec as Gateway from %v", config)
+	}
+
+	var serviceNamespace string
+	var serviceName string
+	var err error
+
+	serviceNamespace, serviceName, err = r.findService(gateway)
+
+	if err != nil {
+		return errors.Wrapf(err, "gateway %s/%s does not specify a valid service", config.Namespace, config.Name)
+	}
+
+	routeMap := make(map[string]v1.Route)
+
+	for _, v := range routes.Items {
+		routeMap[v.Name] = v
+	}
+
+	var result *multierror.Error
+
+	for _, server := range gateway.Servers {
+		for _, host := range server.Hosts {
+			var route *v1.Route
+			var err error
+
+			name := getRouteName(config.Namespace, config.Name, host)
+
+			_, found := routeMap[name]
+
+			if found {
+				route, err = r.updateRoute(config.Meta, host, server.Tls, serviceNamespace, serviceName)
+			} else {
+				route, err = r.createRoute(config.Meta, host, server.Tls, serviceNamespace, serviceName)
+			}
+
+			if err != nil {
+				result = multierror.Append(result, err)
+			} else {
+				delete(routeMap, route.Name)
+			}
+		}
+	}
+
+	for k, v := range routeMap {
+		IORLog.Debugf("clean up route %s for host %s", k, getHost(v))
+		if err := r.deleteRoute(&v); err != nil {
```

Comment:
This line deletes the route if the update of the route fails (in line 406).

Reply:
Updated to remove the `Route` from the current `Route` map even if the update fails. I agree that the deletion should only happen when the user tells us to.